t/storage_zip_audit.t

```diff
@@ -0,0 +1,232 @@
+  }
+
+  describe "HTFeed::StorageZipAudit::PrefixedVersions" => sub {
+    describe "#new" => sub {
+      it "succeeds" => sub {
+        my $storage = prepare_storage('prefixedversions-test');
+        my $audit = audit_for_storage($storage);
+        ok($audit, 'new returns a value');
+        is($audit->{storage_name}, 'prefixedversions-test', 'new returns correct storage name');
+        ok(ref $audit eq 'HTFeed::StorageZipAudit::PrefixedVersions', 'new returns correct class');
+      };
+    };
+  
+    describe "random_object" => sub {
+      it "succeeds" => sub {
+        my $storage = prepare_storage('prefixedversions-test');
+        my $audit = audit_for_storage($storage);
+        my $obj = $audit->random_object();
+        is($obj->{namespace}, 'test', 'random_object returns namespace "test"');
+        is($obj->{objid}, 'test', 'random_object returns objid "test"');
+        is($obj->{version}, $storage->{timestamp}, 'random_object returns timestamp');
+        is($obj->{path}, $storage->audit_path(), 'random_object returns audit path');
+      };
+    };
+
+    describe "#run" => sub {
+      it "succeeds" => sub {
+        my $storage = prepare_storage('prefixedversions-test');
+        my $audit = audit_for_storage($storage);
+        my $errs = $audit->run();
+        is($errs, 0, 'run() returns 0 errors');
+      };
+
+      it "fails when METS checksum is altered" => sub {
+        my $storage = prepare_storage('prefixedversions-test');
+        my $mets_file = $storage->object_path() . '/' . $storage->mets_filename();
+        `sed -i s/2d40d65c1aecd857b3f780e85bc9bd92/2d40d65c1aecd857b3f780e85bc9bd91/g $mets_file`;
+        my $audit = audit_for_storage($storage);
+        my $errs = $audit->run();
+        is($errs, 1, 'run() returns 1 error');
+        my $sql = 'SELECT COUNT(*) FROM feed_audit_detail' . 
+                  ' WHERE namespace = ? AND id = ? AND status = "BadChecksum"';
+        my @res = get_dbh->selectrow_array($sql, undef, 'test', 'test');
+        is($res[0], 1, 'BadChecksum error recorded');
+      };
+
+      it "fails when database checksum is altered" => sub {
+        my $storage = prepare_storage('prefixedversions-test');
+        my $sql = 'UPDATE feed_backups SET saved_md5sum=?' .
+                  ' WHERE namespace=? AND id=? AND version=?';
+        get_dbh->prepare($sql)->execute('deadbeef' x 4, 'test', 'test', $storage->{timestamp});
+        my $audit = audit_for_storage($storage);
+        my $errs = $audit->run();
+        is($errs, 1, 'run() returns 1 error');
+        $sql = 'SELECT COUNT(*) FROM feed_audit_detail' . 
+               ' WHERE namespace = ? AND id = ? AND status = "BadChecksum"';
+        my @res = get_dbh->selectrow_array($sql, undef, 'test', 'test');
+        is($res[0], 1, 'BadChecksum error recorded');
+      };
+    };
+  };
+
+  describe "HTFeed::StorageZipAudit::ObjectStore" => sub {
+    describe "#new" => sub {
```

Comment:
The `Test::Spec` `[shared_examples_for](https://metacpan.org/pod/Test::Spec#shared_examples_for-DESCRIPTION-=%3E-CODE)` feature combined with the `around` hook for setting the `S3_HEAD_OBJECT_RESTORE_DONE` environment variables might be worth looking at for the things shared between groups. For the things that vary between examples (primarily the name of the storage and the storage class) you could look at the `[share %HASH](https://metacpan.org/pod/Test::Spec#share-%HASH)` feature of `Test::Spec`.